### PR TITLE
Various fixes from acceptance testing

### PR DESCRIPTION
**Pull request by danielballan**
See commit messages for details. Changes relate to:

* Handling more edge cases in the test utility ``compare``
* Compatibility with v0 when the Run Stop document is missing
* Handling ragged arrays

### CODE REVIEW

**Comment by danielballan** [comment]
CI failures are random failures.

File: databroker/core.py [hunk 1]
```diff
@@ -1578,11 +1578,19 @@ def _transpose(in_data, keys, field):
...
-            out[k] = dask.array.stack(out[k])
-        else:
-            out[k] = dask.array.array(out[k])
+        try:
+            # compatibility with dask < 2
+            if hasattr(out[k][0], 'shape'):
+                out[k] = dask.array.stack(out[k])
+            else:
+                out[k] = dask.array.array(out[k])
+        except NotImplementedError:
```
[the remark this answers is not in the document]
Comment (danielballan) [reply databroker/core.py:1587]: Interesting. The case that brought me here was ragged arrays, which numpy/dask have no real solution for and thus raise `NotImplementedError`. If we hit the code path you linked, we probably have a lying descriptor and should fix it with a transformer rather than pulling (potentially large) data into memory as a numpy array.

File: databroker/core.py [hunk 2]
```diff
@@ -1578,11 +1578,19 @@ def _transpose(in_data, keys, field):
...
+            # There are data structured that dask auto-chunking cannot handle,
+            # such as an list of list of variable length. For now, let these go
+            # out as plain numpy arrays. In the future we might make them dask
+            # arrays with manual chunks.
+            out[k] = numpy.asarray(out[k])
+            pass
```
Comment (gwbischof) [inline databroker/core.py]: Do we need this pass statement? What is its purpose?

File: databroker/core.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (danielballan) [reply databroker/core.py:1587]: OTOH, it might be clever to add an escape hatch here in case we are overwhelmed by transform-related issues during the ~reply~ deployment. Something like, "If the environment variable `DATABROKER_ARRAY_FALLBACK` is set, catch `ValueError`, warn that things will be slow, and return a numpy array rather than failing." I don't think that should be on by _default_ though.

[the remark this answers is not in the document]
Comment (gwbischof) [reply databroker/core.py:1587]: Won't we hit the ValueError when we call dask.array.stack() with a ragged array?

[the remark this answers is not in the document]
Comment (danielballan) [reply databroker/core.py:1587]: No, numpy treats a ragged array as an array of "objects" (which happen to be lists or variable length) and thus we hit https://github.com/dask/dask/blob/5f61f7f79b861be5ffdbf59950889166628a8fa6/dask/array/core.py#L2532-L2536

File: databroker/core.py (hunk 2 above)
Comment (danielballan) [reply databroker/core.py]: Oops, left over from iterating on this. Pushed 541c93f7 to remove it.

File: databroker/core.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (danielballan) [reply databroker/core.py:1587]: Added emergency fallback in 6a04e03f

File: databroker/tests/test_v2/generic.py [hunk 3]
```diff
@@ -28,26 +28,46 @@ def compare(a, b):
     a_indexed = {}
     b_indexed = {}
     for name, doc in a:
+        if name == 'resource':
+            # Check for an extraneous duplicate key in old documents.
+            if 'id' in doc:
+                assert doc['id'] == doc['uid']
+                doc = doc.copy()
+                doc.pop('id')
         if name == 'datum':
```
Comment (gwbischof) [inline databroker/tests/test_v2/generic.py]: I think this should be an elif

Comment (gwbischof) [inline databroker/tests/test_v2/generic.py:37]: I think this should be an elif

File: databroker/tests/test_v2/generic.py [hunk 4]
```diff
@@ -28,26 +28,46 @@ def compare(a, b):
...
         else:
             a_indexed[(name, doc['uid'])] = doc
     for name, doc in b:
+        if name == 'resource':
+            # Check for an extraneous duplicate key in old documents.
+            if 'id' in doc:
+                assert doc['id'] == doc['uid']
+                doc = doc.copy()
+                doc.pop('id')
         if name == 'datum':
```
Comment (gwbischof) [inline databroker/tests/test_v2/generic.py:51]: I think this should be an elif

File: databroker/tests/test_v2/generic.py (hunk 3 above)
Comment (danielballan) [reply databroker/tests/test_v2/generic.py:37]: Good catch. Pushed.

File: databroker/tests/test_v2/generic.py (hunk 4 above)
Comment (danielballan) [reply databroker/tests/test_v2/generic.py:51]: Ditto.

File: databroker/tests/test_v2/generic.py [hunk 5]
```diff
@@ -28,26 +28,46 @@ def compare(a, b):
     a_indexed = {}
     b_indexed = {}
     for name, doc in a:
-        if name == 'datum':
+        if name == 'resource':
+            # Check for an extraneous duplicate key in old documents.
+            if 'id' in doc:
+                assert doc['id'] == doc['uid']
+                doc = doc.copy()
+                doc.pop('id')
```
Comment (gwbischof) [inline databroker/tests/test_v2/generic.py:36]: The resource isn't being added to a_indexed, same below.

File: databroker/tests/test_v2/generic.py [hunk 6]
```diff
@@ -28,26 +28,46 @@ def compare(a, b):
...
+        elif name == 'datum':
             a_indexed[('datum', doc['datum_id'])] = doc
+        # v0 yields {'_name": 'RunStop'} is the stop doc is missing; v2 yields None.
```
Comment (gwbischof) [inline databroker/tests/test_v2/generic.py]: typo: is -> if. same below

File: databroker/tests/test_v2/generic.py (hunk 3 above)
Comment (danielballan) [reply databroker/tests/test_v2/generic.py:37]: Oops, actually it was intentionally an `if`. First we handle the special case of `resource`; then we go on to handle the various ways that documents contain a unique ID (or not). We want `resource` to hit the `else` path of that second part so that it is added to `a_indexed`.

Comment (danielballan) [reply databroker/tests/test_v2/generic.py:37]: Force-pushed to remove commit 305182ab688ecd8ad8e997f2f997a5ba174c0426 and revert to `if`.

File: databroker/tests/test_v2/generic.py (hunk 4 above)
Comment (danielballan) [reply databroker/tests/test_v2/generic.py:51]: See https://github.com/bluesky/databroker/pull/519#discussion_r380917557

File: databroker/tests/test_v2/generic.py (hunk 5 above)
Comment (danielballan) [reply databroker/tests/test_v2/generic.py:36]: See https://github.com/bluesky/databroker/pull/519#discussion_r380917557

File: databroker/tests/test_v2/generic.py (hunk 6 above)
Comment (danielballan) [reply databroker/tests/test_v2/generic.py]: Fixed.
